lib/fs/fixfs.go

```diff
@@ -0,0 +1,159 @@
+//
+// For implementations, see
+// https://github.com/mirror/newlib-cygwin/blob/fb01286fab9b370c86323f84a46285cfbebfe4ff/winsup/cygwin/path.cc#L435
+// https://github.com/billziss-gh/winfsp/blob/6e3a8f70b2bd958960012447544d492fc6a2f1af/src/shared/ku/posix.c#L1250
+func NewFixingFilesystem(fs Filesystem) Filesystem {
+	return wrapFilesystem(fs, func(underlying Filesystem) Filesystem {
+		return &fixingFilesystem{
+			Filesystem: underlying,
+		}
+	})
+}
+
+func newFixingFilesystem(fs Filesystem) *fixingFilesystem {
+	return &fixingFilesystem{
+		Filesystem: fs,
+	}
+}
+
+func (f *fixingFilesystem) Chmod(name string, mode FileMode) error {
+	return f.Filesystem.Chmod(fixInvalidPath(name), mode)
+}
+
+func (f *fixingFilesystem) Lchown(name string, uid, gid int) error {
+	return f.Filesystem.Lchown(fixInvalidPath(name), uid, gid)
+}
+
+func (f *fixingFilesystem) Chtimes(name string, atime time.Time, mtime time.Time) error {
+	return f.Filesystem.Chtimes(fixInvalidPath(name), atime, mtime)
+}
+
+func (f *fixingFilesystem) Mkdir(name string, perm FileMode) error {
+	return f.Filesystem.Mkdir(fixInvalidPath(name), perm)
+}
+
+func (f *fixingFilesystem) MkdirAll(path string, perm FileMode) error {
+	return f.Filesystem.MkdirAll(fixInvalidPath(path), perm)
+}
+
+func (f *fixingFilesystem) Lstat(name string) (FileInfo, error) {
+	return f.Filesystem.Lstat(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) Remove(name string) error {
+	return f.Filesystem.Remove(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) RemoveAll(name string) error {
+	return f.Filesystem.RemoveAll(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) Rename(oldpath, newpath string) error {
+	return f.Filesystem.Rename(fixInvalidPath(oldpath), fixInvalidPath(newpath))
+}
+
+func (f *fixingFilesystem) Stat(name string) (FileInfo, error) {
+	return f.Filesystem.Stat(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) DirNames(name string) ([]string, error) {
+	dirs, err := f.Filesystem.DirNames(fixInvalidPath(name))
+	if err != nil {
+		return nil, err
+	}
+	for i, dir := range dirs {
+		dirs[i] = revertValidPath(dir)
+	}
+	return dirs, nil
+}
+
+func (f *fixingFilesystem) Open(name string) (File, error) {
+	return f.Filesystem.Open(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) OpenFile(name string, flags int, mode FileMode) (File, error) {
+	return f.Filesystem.OpenFile(fixInvalidPath(name), flags, mode)
+}
+
+func (f *fixingFilesystem) ReadSymlink(name string) (string, error) {
+	return f.Filesystem.ReadSymlink(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) Create(name string) (File, error) {
+	return f.Filesystem.Create(fixInvalidPath(name))
+}
+
+func (f *fixingFilesystem) CreateSymlink(target, name string) error {
+	return f.Filesystem.CreateSymlink(fixInvalidPath(target), fixInvalidPath(name))
+}
+
+type fixFileInfo struct {
+	FileInfo
+}
+
+func (fi *fixFileInfo) Name() string {
+	return fixInvalidPath(fi.FileInfo.Name())
```

Comment:
Isn't this the wrong way around, i.e. the caller should see the "normal" name, without the private unicode bits? That's what you do with the path in `Walk` below.

Reply:
@imsodin Good question, but the caller will see the "normal" ASCII characters as [fixInvalidPath](https://github.com/syncthing/syncthing/blob/0c3a62eaad958e7054aaae68c5a4572e4dd2cff0/lib/fs/fix_windows.go#L38) converts the `\uf0xx` characters to them. It's [revertValidPath](https://github.com/syncthing/syncthing/blob/0c3a62eaad958e7054aaae68c5a4572e4dd2cff0/lib/fs/fix_windows.go#L109) that converts the normal characters back to the Private Use characters.

As @calmh said, these names could be improved upon. Perhaps `encodeNameToPrivate` and `decodeNameFromPrivate`? I am open to suggestions.

Reply:
You're absolutely right. Thanks for catching.